### PR TITLE
New command - Get-DbaDbFileMapping

### LINKED ISSUE #3080: Add a OutFileMapping switch and output to Get-DbaDatabaseFile
### Problem to solve
Add an easy way to build the FileMapping Hashtable for Restore-DbaDatabase from an existing database

For example:
`$FileMapping = Get-DbaDatabaseFile -SqlInstance Instance1 -Database db1 -OutFileMapping`

Would return $FileMapping of type Hashtable like this:
```
$FileMapping = @{
"Datafile1"  = "c:\data\datafile1.mdf";
"Datafile2" = "d:\ArchiveData\File2.mdf";
"LogFile1" = "e:\LogFiles\Log1.ldf"}
```

### PR DESCRIPTION
**Pull request by potatoqualitee**
## Type of Change
 - [x] New feature (non-breaking change, adds functionality, fixes #3080)

### CODE REVIEW

File: functions/Get-DbaDbFileMap.ps1
Comment (andreasjordan) [inline functions/Get-DbaDbFileMap.ps1:51]: I would use [string[]] here to indicate that it's the database name and not a database SMO object.

Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:51]: We use `[object[]]` @andreasjordan because TEPP populates the list of Databases for dynamic validate sets. If you use `string` type this won't be accessible.

File: tests/Get-DbaDbFileMap.Tests.ps1
Comment (wsmelton) [inline tests/Get-DbaDbFileMap.Tests.ps1:13]: We changed the pattern on this test to help on performance, `Get-Command` is very slow to use constantly.

```powershell
    Context "Validate parameters" {
        [array]$params = ([Management.Automation.CommandMetaData]$ExecutionContext.SessionState.InvokeCommand.GetCommand($CommandName, 'Function')).Parameters.Keys
        [object[]]$knownParameters = 'SqlInstance', 'SqlCredential', 'Database', 'InputObject', 'EnableException'

        It "Should only contain our specific parameters" {
            Compare-Object -ReferenceObject $knownParameters -DifferenceObject $params | Should -BeNullOrEmpty
        }
    }
```

File: functions/Get-DbaDbFileMap.ps1
Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:51]: Gives this ability when you use `[object[]` on a parameter where we have the backend setup for TEPP.

image

File: functions/Get-DbaDbFileMap.ps1
Comment (wsmelton) [inline functions/Get-DbaDbFileMap.ps1:83]: Can we call this `FileMapping` since it will contain multiple files and it will match the parameter name we use in the Restore command.

File: functions/Get-DbaDbFileMap.ps1
Comment (andreasjordan) [reply functions/Get-DbaDbFileMap.ps1:51]: Ok, didn't know that. Thanks. Will have a look at the commands I changed in the past.

Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:51]: We can add pretty much anything we can get from SMO or t-sql to list out. Primary one is database. Not the easiest system to deal with and not always consistent in some environments I touch.

File: functions/Get-DbaDbFileMap.ps1
Comment (potatoqualitee) [reply functions/Get-DbaDbFileMap.ps1:83]: ohh i like it, yes

File: functions/Get-DbaDbFileMap.ps1
Comment (potatoqualitee) [reply functions/Get-DbaDbFileMap.ps1:51]: wait for real? i had used object in the past to allow both SMO and text but then went to string[] instead after learning to use `InputObject` better. I'll see if string[] destroys this

Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:51]: It does 😉 because TEPP has to be an object type to return the list to tab completor.

Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:51]: We covered a few initially and at one point I had added in some as I worked on commands, but in that phase of maintainers taking a vacation for a bit 🍹 I never got back to adding support for the main parameters in commands.

File: functions/Get-DbaDbFileMap.ps1
Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:83]: One thing tied to this as well with the name of the command. Should we call it `Get-DbaDbFileMapping` as well? I mean when you think about how it is phrased when we are talking about it or maybe in docs as well, we term it has "you need to check the database mapping" or "what is the mapping of the database files" 🤷🏻

Comment (wsmelton) [reply functions/Get-DbaDbFileMap.ps1:83]: One of those is "Mapping" plural for "Map" in this sense. If we are tying the command directly to the use of the parameter on the Restore command it would make sense to call the command the same as well. Not a hard yes for me, just thinking out loud.

I would also include an update to help on Restore for that parameter and give an example in that using this command so it is found more easily.

Comment (potatoqualitee) [reply functions/Get-DbaDbFileMap.ps1]: Just renamed 👍🏼

File: functions/Get-DbaDbFileMap.ps1
Comment (potatoqualitee) [reply functions/Get-DbaDbFileMap.ps1]: yay, just tested, `[string[]]` does work 💯  updated

**Comment by potatoqualitee** [comment]
Someone let me know when this is good to merge 👍🏼

File: functions/Get-DbaDbFileMapping.ps1
Comment (wsmelton) [inline functions/Get-DbaDbFileMapping.ps1:45]: I would add this example to the restore command as well and then I'm good to merge

**Review by wsmelton — APPROVED** [review]

File: functions/Get-DbaDbFileMapping.ps1
Comment (wsmelton) [reply functions/Get-DbaDbFileMapping.ps1:45]: Might replace the example I think the restore command has for manually building the hashtable object.

Comment (potatoqualitee) [reply functions/Get-DbaDbFileMapping.ps1:45]: I'll add it in addition to -- I think it's important to show ppl how to build the hashtable.

Comment (potatoqualitee) [reply functions/Get-DbaDbFileMapping.ps1:45]: Ty for the approval! Will merge once tests finish.

**Pull request merged**
